src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/WebClientConfiguration.kt

```diff
@@ -61,16 +91,56 @@ class WebClientConfiguration(
     clientService: OAuth2AuthorizedClientService?
   ): OAuth2AuthorizedClientManager? {
     val authorizedClientProvider = OAuth2AuthorizedClientProviderBuilder.builder()
-      .clientCredentials()
+      .clientCredentials {
+        it.accessTokenResponseClient(createRetryingTokenResponseClient())
+      }
       .build()
+
     val authorizedClientManager = AuthorizedClientServiceOAuth2AuthorizedClientManager(
       clientRegistrationRepository,
       clientService
     )
     authorizedClientManager.setAuthorizedClientProvider(authorizedClientProvider)
+
     return authorizedClientManager
   }
 
+  private fun createRetryingTokenResponseClient(): DefaultClientCredentialsTokenResponseClient {
+    val clientCredentialsTokenResponseClient = DefaultClientCredentialsTokenResponseClient()
+
+    val retryCustomizer = RestTemplateCustomizer { restTemplate ->
+      restTemplate.interceptors.add { request, body, execution ->
+
+        val retryTemplate = RetryTemplate()
+        val fixedBackOffPolicy = FixedBackOffPolicy()
+        fixedBackOffPolicy.backOffPeriod = tokenRequestConfig.retryDelayMs
+        retryTemplate.setBackOffPolicy(fixedBackOffPolicy)
+        retryTemplate.setRetryPolicy(SimpleRetryPolicy(tokenRequestConfig.retries))
+        retryTemplate.setListeners(arrayOf(RetryLogger("token request failed; retrying")))
+
+        retryTemplate.execute(
+          RetryCallback<ClientHttpResponse, RestClientException> {
+            execution.execute(request, body)
+          }
+        )
+      }
+    }
+
+    val restTemplate = restTemplateBuilder
+      .setConnectTimeout(Duration.ofMillis(tokenRequestConfig.connectTimeoutMs))
+      .setReadTimeout(Duration.ofMillis(tokenRequestConfig.readTimeoutMs))
+      .customizers(retryCustomizer)
+      .messageConverters(
+        FormHttpMessageConverter(),
+        OAuth2AccessTokenResponseHttpMessageConverter(),
+      )
+      .errorHandler(OAuth2ErrorResponseErrorHandler())
```

Comment:
_nit_ I'm assuming it's not simple to rely on the defaults for these from a pre-configured thing?

Reply:
yeh correct. the docs specify what you need to do to configure the default options 

https://docs.spring.io/spring-security/site/docs/current/reference/html5/#customizing-the-access-token-response-3

in the code, it's in the `DefaultClientCredentialsTokenResponseClient` constructor, and stored in a private `RestOperations`.

```
	public DefaultClientCredentialsTokenResponseClient() {
		RestTemplate restTemplate = new RestTemplate(
				Arrays.asList(new FormHttpMessageConverter(), new OAuth2AccessTokenResponseHttpMessageConverter()));
		restTemplate.setErrorHandler(new OAuth2ErrorResponseErrorHandler());
		this.restOperations = restTemplate;
	}
```